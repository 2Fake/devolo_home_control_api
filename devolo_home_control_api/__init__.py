"""the devolo Home Control API."""
from importlib.metadata import PackageNotFoundError, version

try:
    __version__ = version("devolo_home_control_api")
except PackageNotFoundError:
    # package is not installed - e.g. pulled and run locally
    __version__ = "0.0.0"
