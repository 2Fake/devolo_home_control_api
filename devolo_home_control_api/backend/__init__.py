MESSAGE_TYPES = {
    "devolo.BinarySensor": "_binary_sensor",
    "devolo.BinarySwitch": "_binary_switch",
    "devolo.Blinds": "_multi_level_switch",
    "devolo.DevicesPage": "_inspect_devices",
    "devolo.DewpointSensor": "_multi_level_sensor",
    "devolo.Dimmer": "_multi_level_switch",
    "devolo.Grouping": "_grouping",
    "devolo.HumidityBarValue": "_humidity_bar",
    "devolo.HumidityBarZone": "_humidity_bar",
    "devolo.mprm.gw.GatewayAccessibilityFI": "_gateway_accessible",
    "devolo.Meter": "_meter",
    "devolo.MildewSensor": "_binary_sensor",
    "devolo.MultiLevelSensor": "_multi_level_sensor",
    "devolo.MultiLevelSwitch": "_multi_level_switch",
    "devolo.RemoteControl": "_remote_control",
    "devolo.SirenMultiLevelSwitch": "_multi_level_switch",
    "devolo.ShutterMovementFI": "_binary_sensor",
    "devolo.ValveTemperatureSensor": "_multi_level_sensor",
    "devolo.VoltageMultiLevelSensor": "_multi_level_sensor",
    "devolo.WarningBinaryFI": "_binary_sensor",
    "hdm": "_device_state",
    "acs.hdm": "_automatic_calibration",
    "bas.hdm": "_binary_async",
    "bss.hdm": "_binary_sync",
    "cps.hdm": "_parameter",
    "gds.hdm": "_general_device",
    "lis.hdm": "_led",
    "mas.hdm": "_multilevel_async",
    "mss.hdm": "_multilevel_sync",
    "ps.hdm": "_protection",
    "stmss.hdm": "_multilevel_sync",
    "sts.hdm": "_switch_type",
    "trs.hdm": "_temperature_report",
    "vfs.hdm": "_led"
}
