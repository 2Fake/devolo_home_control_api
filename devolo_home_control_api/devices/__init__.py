"""Devices appearing in a devolo Home Control setup."""
from .gateway import Gateway
from .zwave import Zwave

__all__ = ["Gateway", "Zwave"]
