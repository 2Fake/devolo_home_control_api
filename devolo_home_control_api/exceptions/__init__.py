"""Expections used in this package."""
from .device import SwitchingProtected, WrongElementError
from .gateway import GatewayOfflineError
from .general import WrongCredentialsError, WrongUrlError

__all__ = ["GatewayOfflineError", "SwitchingProtected", "WrongCredentialsError", "WrongElementError", "WrongUrlError"]
