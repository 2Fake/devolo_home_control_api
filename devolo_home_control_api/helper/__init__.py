"""Helper functions used in the package."""
from .names import camel_case_to_snake_case
from .uid import (
    get_device_type_from_element_uid,
    get_device_uid_from_element_uid,
    get_device_uid_from_setting_uid,
    get_home_id_from_device_uid,
    get_sub_device_uid_from_element_uid,
)

__all__ = [
    "camel_case_to_snake_case",
    "get_device_type_from_element_uid",
    "get_device_uid_from_element_uid",
    "get_device_uid_from_setting_uid",
    "get_home_id_from_device_uid",
    "get_sub_device_uid_from_element_uid",
]
