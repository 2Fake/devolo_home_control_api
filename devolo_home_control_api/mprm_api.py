import requests
import json
import websocket
import time
import threading

from mydevolo_api import Mydevolo


class MprmRestApi:
    def __init__(self, user, password, gateway_serial, mydevolo_url='https://www.mydevolo.com', mprm_url='https://homecontrol.mydevolo.com'):
        mydevolo = Mydevolo(user=user, password=password, url=mydevolo_url)
        self._mprm_url = mprm_url
        self._gateway_serial = gateway_serial
        self._headers = {'content-type': 'application/json'}
        uuid = mydevolo.get_uuid()

        self.rpc_url = self._mprm_url + "/remote/json-rpc"

        # Create a session
        self.session = requests.Session()
        full_url = requests.get(mydevolo.get_url() + "/v1/users/" + uuid + "/hc/gateways/" + self._gateway_serial + "/fullURL", auth=(user, password), headers=self._headers).json()['url']
        self.session.get(full_url)

        # create a dict with UIDs --> names
        self._element_uid_dict = {}
        # create a dict with names --> UIDs
        self._devices = {}
        self._groups = {}
        self._schedules = {}
        self._scenes = {}
        self._notifications = {}
        self._rules = {}
        self.update_devices()
        self.update_groups()
        self.update_schedules()
        self.update_scenes()
        self.update_notifications()
        self.update_rules()

        element_uid_list = []
        for uid, device_name in self._element_uid_dict.items():
            for element_uid in device_name:
                if element_uid.startswith('devolo.Meter'):
                    element_uid_list.append(f'current_consumption_{element_uid}')
                elif element_uid.startswith('devolo.BinarySwitch'):
                    element_uid_list.append(f'binary_state_{element_uid}')
        self._pub = Publisher(element_uid_list)
        print(element_uid_list)
        self.register_pub()

    def get_consumption(self, uid, consumption_type='current'):
        if consumption_type not in ['current', 'total']:
            raise ValueError('Unknown consumption type. "current" and "total" are valid consumption types.')
        # TODO: Prepare for more meter items as one
        return self._element_uid_dict.get(uid).get(f'devolo.Meter:{uid}').get(f'{consumption_type}_consumption')

    def update_binary_switch_state(self, uid, value=None):
        element_uid = 'devolo.BinarySwitch:' + uid
        if value is None:
            r = self._extract_data_from_element_uid(element_uid)
            self._element_uid_dict[uid][element_uid]['state'] = True if r['properties']['state'] == 1 else False
        else:
            self._element_uid_dict[uid][element_uid]['binary_switch']['state'] = value
            self._pub.dispatch(f'state_{self._element_uid_dict[element_uid]}', value)

    def update_consumption(self, uid, consumption, value=None):
        if consumption not in ['current', 'total']:
            raise ValueError('Consumption value is not valid. Only "current" and "total are allowed!')
        if value is not None:
            raise ValueError('Value is not allowed here.')
        element_uid = 'devolo.Meter:' + uid
        r = self._extract_data_from_element_uid(element_uid)
        value = r['properties']['currentValue'] if consumption == 'currentValue' else r['properties']['totalValue']
        self._element_uid_dict[uid][element_uid]['current_consumption'] = value

    def get_binary_switch_state(self, uid):
        return self._element_uid_dict.get(uid).get(f'devolo.BinarySwitch:{uid}').get('state')

    def get_current_consumption(self, uid):
        try:
            return self._element_uid_dict.get(uid).get(f'devolo.Meter:{uid}').get('current_consumption')
        except AttributeError:
            # TODO 1D Relay does not have a consumption. We should do a better error handling here.
            return None

    def set_binary_switch_state(self, uid: str, state: bool):
        # TODO Change to UUID
        """
        Set the binary switch to the desired state
        :param uid: uid
        :param state: Desired state of the binary switch of the device
        :return:
        """
        self.set_binary_switch(element_uid=f'devolo.BinarySwitch:{uid}', state=state)

    def register_pub(self):
        for uid in self._element_uid_dict:
            for element_uid in self._element_uid_dict[uid]:
                if element_uid.startswith('devolo.Meter'):
                    self._element_uid_dict[uid][element_uid]['subscriber'] = Subscriber(uid)
                    self._pub.register(f'current_consumption_{element_uid}', self._element_uid_dict[uid][element_uid]['subscriber'])
                elif element_uid.startswith('devolo.BinarySwitch'):
                    self._element_uid_dict[uid][element_uid]['subscriber'] = Subscriber(uid)
                    self._pub.register(f'binary_state_{element_uid}', self._element_uid_dict[uid][element_uid]['subscriber'])

    def update_devices(self):
        # TODO: Add http, hue, powermeter
        data = {'jsonrpc': '2.0',
                'id': 10,
                'method': 'FIM/getFunctionalItems',
                'params': [["devolo.DevicesPage"], 0]}
        r = self.session.post(self.rpc_url, data=json.dumps(data), headers=self._headers)
        self._devices = {}
        self._element_uid_dict = {}
        for item in r.json()['result']['items']:
            all_devices_list = item['properties']['deviceUIDs']
            for device in all_devices_list:
                name, element_uids = self._get_name_and_element_uids(uid=device)
                self._element_uid_dict[device] = {}
                for uid in element_uids:
                    self._element_uid_dict[device][uid] = {}
                    self._element_uid_dict[device]['name'] = name

    def get_devices(self):
        return self._devices

    def get_binary_switch_devices(self):
        devices = []
        for uid, device in self._element_uid_dict.items():
            [devices.append(device) for element_uid in self._element_uid_dict.get(uid) if element_uid.startswith('devolo.BinarySwitch')]
        return devices

    def update_groups(self):
        data = {'jsonrpc': '2.0',
                'id': 10,
                'method': 'FIM/getFunctionalItems',
                'params': [["devolo.Grouping"], 0]}
        r = self.session.post(self.rpc_url, data=json.dumps(data), headers=self._headers)
        self._groups = {}
        for item in r.json()['result']['items']:
            all_groups_list = item['properties']['smartGroupWidgetUIDs']
            for group in all_groups_list:
                name, elementUIDs = self._get_name_and_element_uids(uid=group)
                self._groups[name] = elementUIDs

    def update_schedules(self):
        data = {'jsonrpc': '2.0',
                'id': 10,
                'method': 'FIM/getFunctionalItems',
                'params': [["devolo.Schedules"], 0]}
        r = self.session.post(self.rpc_url, data=json.dumps(data), headers=self._headers)
        self._schedules = {}
        for item in r.json()['result']['items']:
            all_schedules_list = item['properties']['scheduleUIDs']
            for schedule in all_schedules_list:
                name, elementUIDs = self._get_name_and_element_uids(uid=schedule)
                self._schedules[name] = elementUIDs

    def update_notifications(self):
        data = {'jsonrpc': '2.0',
                'id': 10,
                'method': 'FIM/getFunctionalItems',
                'params': [["devolo.Messages"], 0]}
        r = self.session.post(self.rpc_url, data=json.dumps(data), headers=self._headers)
        self._notifications = {}
        for item in r.json()['result']['items']:
            all_notifications_list = item['properties']['customMessageUIDs']
            for notification in all_notifications_list:
                name, elementUIDs = self._get_name_and_element_uids(uid=notification)
                self._notifications[name] = elementUIDs

    def update_rules(self):
        data = {'jsonrpc': '2.0',
                'id': 10,
                'method': 'FIM/getFunctionalItems',
                'params': [["devolo.Services"], 0]}
        r = self.session.post(self.rpc_url, data=json.dumps(data), headers=self._headers)
        self._rules = {}
        for item in r.json()['result']['items']:
            all_rules_list = item['properties']['serviceUIDs']
            for rule in all_rules_list:
                name, elementUIDs = self._get_name_and_element_uids(uid=rule)
                self._rules[name] = elementUIDs

    def update_scenes(self):
        data = {'jsonrpc': '2.0',
                'id': 10,
                'method': 'FIM/getFunctionalItems',
                'params': [["devolo.Scene"], 0]}
        r = self.session.post(self.rpc_url, data=json.dumps(data), headers=self._headers)
        self._scenes = {}
        for item in r.json()['result']['items']:
            all_scenes_list = item['properties']['sceneUIDs']
            for scene in all_scenes_list:
                name, elementUIDs = self._get_name_and_element_uids(uid=scene)
                self._scenes[name] = elementUIDs

    def _get_name_and_element_uids(self, uid):
        data = {'jsonrpc': '2.0',
                'id': 11,
                'method': 'FIM/getFunctionalItems',
                'params': [[f"{uid}"], 0]}
        r = self.session.post(self.rpc_url, data=json.dumps(data), headers=self._headers)
        for x in r.json()["result"]["items"]:
            return x['properties']['itemName'], x['properties']["elementUIDs"]

    def _extract_data_from_element_uid(self, element_uid):
        data = {'jsonrpc': '2.0',
                'id': 11,
                'method': 'FIM/getFunctionalItems',
                'params': [[f"{element_uid}"], 0]}
        r = self.session.post(self.rpc_url, data=json.dumps(data), headers=self._headers)
        # TODO: Catch error!
        return r.json()['result']['items'][0]

    def set_binary_switch(self, element_uid, state: bool):
        data = {'jsonrpc': '2.0',
                'id': 11,
                'method': 'FIM/invokeOperation',
                'params': [f"{element_uid}", "turnOn" if state else "turnOff", []]}
        r = self.session.post(self.rpc_url, data=json.dumps(data), headers=self._headers)
        # TODO: Catch errors!

    def get_element_uid(self, element_uid):
        """
        Return the FIM UID of the given element uid
        :param element_uid:
        :return: FIM UID as string
        """
        return element_uid.split(":", 1)


class MprmWebSocket(MprmRestApi):
    def __init__(self, user, password, gateway_serial, mydevolo_url='https://www.mydevolo.com', mprm_url='https://homecontrol.mydevolo.com'):
        super().__init__(user, password, gateway_serial, mydevolo_url, mprm_url)
        self._ws = None

    def on_open(self):
        def run(*args):
            # TODO: replace by logger
            print('Starting web socket connection')
            while True:
                time.sleep(1)
            time.sleep(1)
            self._ws.close()
            # TODO: replace by logger
            print("thread terminating...")

        threading.Thread(target=run).start()

    def on_message(self, message):
        message = json.loads(message)
        if message['properties']['uid'].startswith('devolo.Meter'):
            # TODO: distinguish between current and total value
            self.update_consumption(element_uid=message.get("properties").get("uid"), consumption="current", value=message.get('properties').get('property.value.new'))
        elif message['properties']['uid'].startswith('devolo.BinarySwitch') and message['properties']['property.name'] == 'state':
            self.update_binary_switch_state(element_uid=message.get("properties").get("uid"), value=True if message.get('properties').get('property.value.new') == 1 else False)
        else:
            # Unknown messages shall be ignored
            pass

    def on_error(self, error):
        # TODO: replace by logger
        print(error)

    def on_close(self):
        # TODO: replace by logger
        print("### closed ###")

    def web_socket_connection(self, cookies: dict):
        import websocket  # TODO: Find out why we nee the import here. Otherwise it throws an error --> AttributeError: 'function' object has no attribute 'WebSocketApp'
        ws_url = self._mprm_url.replace("https://", "wss://").replace("http://", "ws://")
        cookie = "; ".join([str(name)+"="+str(value) for name, value in cookies.items()])
        ws_url = f"{ws_url}/remote/events/?topics=com/prosyst/mbs/services/fim/FunctionalItemEvent/PROPERTY_CHANGED,com/prosyst/mbs/services/fim/FunctionalItemEvent/UNREGISTERED&filter=(|(GW_ID={self._gateway_serial})(!(GW_ID=*)))"
        self._ws = websocket.WebSocketApp(ws_url,
                                          cookie=cookie,
                                          on_open=self.on_open,
                                          on_message=self.on_message,
                                          on_error=self.on_error,
                                          on_close=self.on_close)
        self._ws.run_forever()

    def update_consumption(self, element_uid, consumption, value=None):
        if consumption not in ['current', 'total']:
            raise ValueError('Consumption value is not valid. Only "current" and "total are allowed!')
        if value is None:
            raise ValueError('Got value==None. This is impossible')
        uid = element_uid.split(":", 1)[1].split("#")[0]
        self._element_uid_dict[uid][element_uid][f'{consumption}_consumption'] = value
        self._pub.dispatch(f'current_consumption_{element_uid}', value)


class Publisher:
    def __init__(self, events):
        # maps event names to subscribers
        # str -> dict
        self.events = {event: dict()
                       for event in events}

    def get_subscribers(self, event):
        return self.events[event]

    def register(self, event, who, callback=None):
        if callback == None:
            callback = getattr(who, 'update')
        self.get_subscribers(event)[who] = callback

    def unregister(self, event, who):
        del self.get_subscribers(event)[who]

    def dispatch(self, event, message):
        for subscriber, callback in self.get_subscribers(event).items():
            callback(message)


class Subscriber:
    def __init__(self, name):
        self.name = name

    def update(self, message):
        print('{} got message "{}"'.format(self.name, message))


if __name__ == "__main__":

    user = "username"
    password = "password"
    mydevolo = Mydevolo(user=user, password=password, url='https://dcloud-test.devolo.net')
    uuid = mydevolo.get_uuid()
    gateways_serials = mydevolo.get_gateway_serials()
    api = MprmRestApi(user=user,
                      password=password,
                      mydevolo_url=mydevolo.get_url(),
                      mprm_url="https://mprm-test.devolo.net",
                      gateway_serial="1406126500001876")
    # print(api.get_binary_switch_devices())

    # state = api.update_binary_switch_state(device_name=device_name)
    # print(f'state: {state}')
    # api.set_binary_switch_state(device_name=device_name, state=not state)
    # sleep(2)
    # state = api.get_binary_switch_state(device_name=device_name)
    # print(f'state: {state}')

    def websocket(*args):
        mprm_websocket = MprmWebSocket(user=user,
                                       password=password,
                                       mydevolo_url=mydevolo.get_url(),
                                       mprm_url="https://mprm-test.devolo.net",
                                       gateway_serial="1406126500001876")
        mprm_websocket.web_socket_connection(cookies=api.session.cookies)

    threading.Thread(target=websocket).start()
    # while True:
    #     time.sleep(1)
    #     print(f'Binary Switch: {api.get_binary_switch_state(device_name=device_name)}')
    #     print(f'Consumption: {api.get_current_consumption(device_name=device_name)}')


