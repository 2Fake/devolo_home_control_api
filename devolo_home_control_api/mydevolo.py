import logging

import requests

from . import __version__
from .exceptions.gateway import GatewayOfflineError
from .exceptions.general import WrongCredentialsError, WrongUrlError


class Mydevolo:
    """
    The Mydevolo object handles calls to the my devolo API v1 as singleton. It does not cover all API calls, just
    those requested up to now. All calls are done in a user context, so you need to provide credentials of that user.
    """

    __instance = None

    @classmethod
    def get_instance(cls):
        if cls.__instance is None:
            raise SyntaxError(f"Please init {cls.__name__}() once to establish a connection to my devolo.")
        return cls.__instance

    @classmethod
    def del_instance(cls):
        cls.__instance = None


    def __init__(self):
        if self.__class__.__instance is not None:
            raise SyntaxError(f"Please use {self.__class__.__name__}.get_instance() and reuse the connection to my devolo.")

        self._logger = logging.getLogger(self.__class__.__name__)
        self._user = None
        self._password = None
        self._uuid = None
        self._gateway_ids = []

        self.url = "https://www.mydevolo.com"

        self.__class__.__instance = self


    @property
    def user(self) -> str:
        """ The user (also known as my devolo ID) is used for basic authentication. """
        return self._user

    @user.setter
    def user(self, user: str):
        """ Invalidate uuid and gateway IDs on user name change. """
        self._user = user
        self._uuid = None
        self._gateway_ids = []

    @property
    def password(self) -> str:
        """ The password is used for basic authentication. """
        return self._password

    @password.setter
    def password(self, password: str):
        """ Invalidate uuid and gateway IDs on password change. """
        self._password = password
        self._uuid = None
        self._gateway_ids = []


    def credentials_valid(self) -> bool:
        """ Check if current credentials are valid. """
        try:
            self.uuid()
            return True
        except WrongCredentialsError:
            return False

    def get_gateway_ids(self) -> list:
        """ Get all gateway IDs. """
        if not self._gateway_ids:
            self._logger.debug("Getting list of gateways")
            items = self._call(f"{self.url}/v1/users/{self.uuid()}/hc/gateways/status").get("items")
            for gateway in items:
                self._gateway_ids.append(gateway.get("gatewayId"))
                self._logger.debug(f'Adding {gateway.get("gatewayId")} to list of gateways.')
            if len(self._gateway_ids) == 0:
                self._logger.error("Could not get gateway list. No Gateway attached to account?")
                raise IndexError("No gateways")
        return self._gateway_ids

    def get_gateway(self, gateway_id: str) -> dict:
        """
        Get gateway details like name, local passkey and other.

        :param gateway_id: Gateway ID
        :return: Gateway object
        """
        self._logger.debug(f"Getting details for gateway {gateway_id}")
        details = {}
        try:
            details = self._call(f"{self.url}/v1/users/{self.uuid()}/hc/gateways/{gateway_id}")
        except WrongUrlError:
            self._logger.error("Could not get full URL. Wrong gateway ID used?")
            raise
        return details

    def get_full_url(self, gateway_id: str) -> str:
        """
        Get gateway's portal URL.

        :param gateway_id: Gateway ID
        :return: URL to access the gateway's portal.
        """
        self._logger.debug("Getting full URL of gateway.")
        return self._call(f"{self.url}/v1/users/{self.uuid()}/hc/gateways/{gateway_id}/fullURL").get("url")

    def get_zwave_products(self, manufacturer: str, product_type: str, product: str) -> dict:
        """
        Get information about a Z-Wave device.

        :param manufacturer: The manufacturer ID in hex.
        :param product_type: The product type ID in hex.
        :param product: The product ID in hex.
        :return: All known product information.
        """
        self._logger.debug(f"Getting information for {manufacturer}/{product_type}/{product}")
        device_info = {}
        try:
            device_info = self._call(f"{self.url}/v1/zwave/products/{manufacturer}/{product_type}/{product}")
        except WrongUrlError:
            # At some devices no device information are returned
            self._logger.debug("No device info found")
            device_info = {
                "brand": "devolo" if manufacturer == "0x0175" else "Unknown",
                "deviceType": "Unknown",
                "genericDeviceClass": "Unknown",
                "href": f"{self.url}/v1/zwave/products/{manufacturer}/{product_type}/{product}",
                "identifier": "Unknown",
                "isZWavePlus": False,
                "manufacturerId": manufacturer,
                "name": "Unknown",
                "productId": product,
                "productTypeId": product_type,
                "specificDeviceClass": "Unknown",
                "zwaveVersion": "Unknown"
            }
        return device_info

    def maintenance(self) -> bool:
        """ If devolo Home Control is in maintenance, there is not much we can do via cloud. """
        state = self._call(f"{self.url}/v1/hc/maintenance").get("state")
        if state == "on":
            return False
        else:
            self._logger.warning("devolo Home Control is in maintenance mode.")
            return True

    def uuid(self) -> str:
        """ The uuid is a central attribute in my devolo. Most URLs in the user's context contain it. """
        if self._uuid is None:
            self._logger.debug("Getting UUID")
            self._uuid = self._call(f"{self.url}/v1/users/uuid").get("uuid")
        return self._uuid


    def _call(self, url: str) -> dict:
        """ Make a call to any entry point with the user's context. """
        headers = {"content-type": "application/json",
                   "User-Agent": f"devolo_home_control_api/{__version__}"}
        responds = requests.get(url,
                                auth=(self._user, self._password),
                                headers=headers,
                                timeout=60)

        if responds.status_code == requests.codes.forbidden:
            self._logger.error("Could not get full URL. Wrong username or password?")
            raise WrongCredentialsError("Wrong username or password.")
        if responds.status_code == requests.codes.not_found:
            raise WrongUrlError(f"Wrong URL: {url}")
        if responds.status_code == requests.codes.service_unavailable:
            self._logger.error("The requested gateway seems to be offline.")
            raise GatewayOfflineError("Gateway offline.")
        return responds.json()
