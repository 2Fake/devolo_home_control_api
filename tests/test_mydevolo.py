import pytest

from devolo_home_control_api.mydevolo import Mydevolo, GatewayOfflineError, WrongUrlError, WrongCredentialsError


class TestMydevolo:
    @pytest.mark.usefixtures("mock_mydevolo__call")
    def test_credentials_valid(self, mydevolo):
        assert mydevolo.credentials_valid()

    @pytest.mark.usefixtures("mock_mydevolo__call_raise_WrongCredentialsError")
    def test_credentials_invalid(self, mydevolo):
        assert not mydevolo.credentials_valid()

    @pytest.mark.usefixtures("mock_mydevolo__call")
    def test_gateway_ids(self, mydevolo):
        assert mydevolo.get_gateway_ids() == [self.gateway.get("id")]

    @pytest.mark.usefixtures("mock_mydevolo__call")
    def test_gateway_ids_empty(self, mydevolo):
        with pytest.raises(IndexError):
            mydevolo.get_gateway_ids()

    @pytest.mark.usefixtures("mock_mydevolo__call")
    def test_get_full_url(self, mydevolo):
        full_url = mydevolo.get_full_url(self.gateway.get("id"))
        assert full_url == self.gateway.get("full_url")

    @pytest.mark.usefixtures("mock_mydevolo__call")
    def test_get_gateway(self, mydevolo):
        details = mydevolo.get_gateway(self.gateway.get("id"))
        assert details.get("gatewayId") == self.gateway.get("id")

    @pytest.mark.usefixtures("mock_mydevolo__call_raise_WrongUrlError")
    def test_get_gateway_invalid(self, mydevolo):
        with pytest.raises(WrongUrlError):
            mydevolo.get_gateway(self.gateway.get("id"))

    @pytest.mark.usefixtures("mock_mydevolo__call")
    def test_get_zwave_products(self, mydevolo):
        product = mydevolo.get_zwave_products(manufacturer="0x0060", product_type="0x0001", product="0x000")
        assert product.get("name") == "Everspring PIR Sensor SP814"

    @pytest.mark.usefixtures("mock_mydevolo__call_raise_WrongUrlError")
    def test_get_zwave_products_invalid(self, mydevolo):
        device_infos = mydevolo.get_zwave_products(manufacturer="0x0070", product_type="0x0001", product="0x000")
        assert device_infos.get("name") == "Unknown"

    @pytest.mark.usefixtures("mock_mydevolo__call")
    def test_maintenance_on(self, mydevolo):
        assert not mydevolo.maintenance()

    @pytest.mark.usefixtures("mock_mydevolo__call")
    def test_maintenance_off(self, mydevolo):
        assert mydevolo.maintenance()

    def test_set_password(self, mydevolo):
        mydevolo._gateway_ids = [self.gateway.get("id")]

        mydevolo.password = self.user.get("password")

        assert mydevolo._uuid is None
        assert mydevolo._gateway_ids == []

    def test_set_user(self, mydevolo):
        mydevolo._gateway_ids = [self.gateway.get("id")]

        mydevolo.user = self.user.get("username")

        assert mydevolo._uuid is None
        assert mydevolo._gateway_ids == []

    def test_get_user(self, mydevolo):
        mydevolo.user = self.user.get("username")
        assert mydevolo.user == self.user.get("username")

    def test_get_password(self, mydevolo):
        mydevolo.password = self.user.get("password")
        assert mydevolo.password == self.user.get("password")

    def test_singleton_mydevolo(self):
        with pytest.raises(SyntaxError):
            Mydevolo.get_instance()

        first = Mydevolo()
        with pytest.raises(SyntaxError):
            Mydevolo()

        second = Mydevolo.get_instance()
        assert first is second
        Mydevolo.del_instance()

    @pytest.mark.usefixtures("mock_mydevolo__call")
    def test_uuid(self, mydevolo):
        mydevolo._uuid = None
        assert mydevolo.uuid() == self.user.get("uuid")

    @pytest.mark.usefixtures("mock_response_wrong_credentials_error")
    def test_call_WrongCredentialsError(self):
        mydevolo = Mydevolo()
        with pytest.raises(WrongCredentialsError):
            mydevolo._call("test")
        Mydevolo.del_instance()

    @pytest.mark.usefixtures("mock_response_wrong_url_error")
    def test_call_WrongUrlError(self):
        mydevolo = Mydevolo()
        with pytest.raises(WrongUrlError):
            mydevolo._call("test")
        Mydevolo.del_instance()

    @pytest.mark.usefixtures("mock_response_gateway_offline")
    def test_call_GatewayOfflineError(self, mydevolo):
        with pytest.raises(GatewayOfflineError):
            mydevolo._call("test")

    @pytest.mark.usefixtures("mock_response_valid")
    def test_call_valid(self, mydevolo):
        assert mydevolo._call("test").get("response") == "response"
