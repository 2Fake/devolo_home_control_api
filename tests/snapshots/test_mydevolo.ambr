# serializer version: 1
# name: test_get_gateway
  dict({
    'externalAccess': True,
    'firmwareVersion': '8.0.45_2016-11-17',
    'gatewayId': '1409301750000598',
    'hcuPartner': 'string',
    'href': 'https://mydevolo.com/v1/users/535512AB-165D-11E7-A4E2-000C29D76CCA/hc/gateways/1409301750000598',
    'invitations': dict({
      'href': 'string',
    }),
    'localPasskey': '6b96ad097aa389209f1ceeaed6fe7029',
    'location': dict({
    }),
    'name': 'DVT Lab 1409301750000598',
    'partner': 'string',
    'role': 'user',
    'sceneActionJournal': dict({
      'href': 'string',
    }),
    'sceneActions': dict({
      'href': 'string',
    }),
    'scenes': dict({
      'href': 'string',
    }),
    'state': 'devolo.hc_gateway.state.idle',
    'status': 'devolo.hc_gateway.status.online',
    'type': 'M',
    'users': dict({
      'href': 'string',
    }),
  })
# ---
# name: test_get_gateway_ids
  list([
    '1409301750000598',
  ])
# ---
# name: test_get_uuid
  '535512AB-165D-11E7-A4E2-000C29D76CCA'
# ---
# name: test_get_zwave_products
  dict({
    'brand': 'Everspring',
    'deviceType': 'Door Lock Keypad',
    'genericDeviceClass': 'Entry Control',
    'href': 'https://mydevolo.com/v1/zwave/products/0x0060/0x0001/0x0002',
    'identifier': 'SP814-US',
    'isZWavePlus': True,
    'manufacturerId': '0x0060',
    'name': 'Everspring PIR Sensor SP814',
    'productId': '0x0002',
    'productTypeId': '0x0001',
    'specificDeviceClass': 'Secure Keypad Door Lock',
    'zwaveVersion': '6.51.07',
  })
# ---
# name: test_get_zwave_products.1
  dict({
    'brand': 'Unknown',
    'deviceType': 'Unknown',
    'genericDeviceClass': 'Unknown',
    'href': 'https://www.mydevolo.com/v1/zwave/products/0x1060/0x0001/0x0002',
    'identifier': 'Unknown',
    'isZWavePlus': False,
    'manufacturerId': '0x1060',
    'name': 'Unknown',
    'productId': '0x0002',
    'productTypeId': '0x0001',
    'specificDeviceClass': 'Unknown',
    'zwaveVersion': 'Unknown',
  })
# ---
